daemon/api_sideload_n_try.go

```diff
@@ -218,39 +219,153 @@ func sideloadOrTrySnap(ctx context.Context, c *Command, body io.ReadCloser, boun
 
 // sideloadedInfo contains information from a bunch of sideloaded snaps
 type sideloadedInfo struct {
-	sideInfos                  []*snap.SideInfo
-	names, origPaths, tmpPaths []string
+	snaps      []sideloadSnapInfo
+	components []sideloadComponentInfo
 }
 
-func sideloadInfo(st *state.State, snapFiles []*uploadedSnap, flags sideloadFlags) (*sideloadedInfo, *apiError) {
+type sideloadSnapInfo struct {
+	sideInfo   *snap.SideInfo
+	components []sideloadComponentInfo
+	origPath   string
+	tmpPath    string
+}
+
+type sideloadComponentInfo struct {
+	sideInfo *snap.ComponentSideInfo
+	origPath string
+	tmpPath  string
+}
+
+func sideloadInfo(st *state.State, uploads []*uploadedContainer, flags sideloadFlags) (*sideloadedInfo, *apiError) {
 	deviceCtx, err := snapstate.DevicePastSeeding(st, nil)
 	if err != nil {
 		return nil, InternalError(err.Error())
 	}
 
-	names := make([]string, len(snapFiles))
-	origPaths := make([]string, len(snapFiles))
-	tmpPaths := make([]string, len(snapFiles))
-	sideInfos := make([]*snap.SideInfo, len(snapFiles))
+	var components []sideloadComponentInfo
+	var snaps []sideloadSnapInfo
+	for _, upload := range uploads {
+		si, snapErr := readSideInfo(st, upload.tmpPath, upload.filename, flags, deviceCtx.Model())
+		if snapErr != nil {
+			if !flags.dangerousOK {
+				// TODO:COMPS: read assertions for components
+				return nil, snapErr
+			}
+
+			ci, err := readComponentInfoFromCont(upload.tmpPath, nil)
+			if err != nil {
+				logger.Noticef("cannot sideload as a snap: %v", snapErr)
+				logger.Noticef("cannot sideload as a component: %v", err)
 
-	for i, snapFile := range snapFiles {
-		si, apiError := readSideInfo(st, snapFile.tmpPath, snapFile.filename, flags, deviceCtx.Model())
-		if apiError != nil {
-			return nil, apiError
+				// note that here we forward the error from reading the snap
+				// file, rather than the component file. this is consistent with
+				// what we do when installing one component from file. maybe
+				// something to change?
+				return nil, snapErr
+			}
+
+			components = append(components, sideloadComponentInfo{
+				sideInfo: &snap.ComponentSideInfo{
+					Component: ci.Component,
+					Revision:  snap.Revision{},
+				},
+				origPath: upload.filename,
+				tmpPath:  upload.tmpPath,
+			})
+			continue
 		}
 
-		sideInfos[i] = si
-		names[i] = si.RealName
-		origPaths[i] = snapFile.filename
-		tmpPaths[i] = snapFile.tmpPath
+		snaps = append(snaps, sideloadSnapInfo{
+			sideInfo: si,
+			origPath: upload.filename,
+			tmpPath:  upload.tmpPath,
+		})
+	}
+
+	snapByName := func(name string) (*sideloadSnapInfo, bool) {
+		for i := range snaps {
+			if snaps[i].sideInfo.RealName == name {
+				return &snaps[i], true
+			}
+		}
+		return nil, false
```

Comment:
wouldn't it be easier to create a map snapsByName parallel to the snaps slice above?

Reply:
We need to mutate the elements returned by snapByName, and storing pointers to elements in the `snaps` slice is potentially incorrect if the slice gets reallocated.

I'll add a comment pointing out why we do it like this.